crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -311,63 +317,47 @@ where
             .collect()
     }
 
-    /// Read a lock for a specific (transaction, shared object) pair.
-    pub fn get_all_shared_locks(
+    /// Read shared object locks / versions for a specific transaction.
+    pub fn get_shared_locks(
         &self,
         transaction_digest: &TransactionDigest,
     ) -> Result<Vec<(ObjectID, SequenceNumber)>, SuiError> {
         Ok(self
             .tables
-            .assigned_object_versions
-            .iter()
-            .skip_to(&(*transaction_digest, ObjectID::ZERO))?
-            .take_while(|((tx, _objid), _ver)| tx == transaction_digest)
-            .map(|((_tx, objid), ver)| (objid, ver))
-            .collect())
+            .assigned_shared_object_versions
+            .get(transaction_digest)?
```

Comment:
Another optimization we could've tried is - provide an upper bound on the iterate through read options `ReadOptions.iterate_upper_bound` (i.e. tx_digest, max_object_id). By setting this option, RocksDB doesn't have to find the next key after the upper bound and can avoid a bunch of comparisons. And used a single delete range instead of single deletes on every objectid for a tx digest (which would have reduced the number of tombstones). But this refactor is better overall as it completely avoids an iter()

Reply:
I'm thinking of adding an iter_prefix() API to typed store later, to avoid the inefficient `iter().skip_to(...).take_while(...)`. The `iterate_upper_bound` setting could be useful there.